File: sdm_cpp/speed_test.cpp
#include "kanerva_sdm.h"
#include <iostream>
#include <iomanip>
#include <vector>
#include <random>
#include <chrono>

// Helper function to generate random binary vector.
std::vector<int> generate_random_vector(int size, std::mt19937& rng) {
    std::uniform_int_distribution<int> dist(0, 1);
    
    std::vector<int> vec(size);
    for (int i = 0; i < size; ++i) {
        vec[i] = dist(rng);
    }
    return vec;
}

int main() {
    KanervaSDM sdm(100, 100, 10000, 37);
    
    const int num_memories = 10000;
    std::mt19937 rng(42);
    
    // Create vectors for storing addresses and memories.
    std::vector<std::vector<int>> addresses(num_memories);
    std::vector<std::vector<int>> memories(num_memories);
    std::vector<std::vector<int>> recalled_memories(num_memories);
    
    // Start timing
    auto start = std::chrono::high_resolution_clock::now();
    
    // Write all memories.
    for (int i = 0; i < num_memories; ++i) {
        std::vector<int> address = generate_random_vector(sdm.get_address_dimension(), rng);
        std::vector<int> memory = generate_random_vector(sdm.get_memory_dimension(), rng);
        addresses[i] = address;
        memories[i] = memory;
        sdm.write(address, memory);
    }
    
    // Read all memories.
    for (int i = 0; i < num_memories; ++i) {
        std::vector<int> address = addresses[i];
        std::vector<int> recalled_memory = sdm.read(address);
        recalled_memories[i] = recalled_memory;
    }
    
    // End timing
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    std::cout << std::fixed << std::setprecision(5);
    std::cout << "Elapsed time: " << duration.count() / 1000000.0 << "s" << std::endl;
    std::cout.flush();  // Force flush
    
    std::cout << "\nPress Enter to exit..." << std::endl;
    std::cin.get();
    
    return 0;
}

File: sdm_cpp/speed_test.py
from KanervaSDM import KanervaSDM
import numpy as np
import time 
from tqdm import tqdm
import os

SDM = KanervaSDM(
    address_dimension=100,
    memory_dimension=100, 
    num_locations=10000, 
    activation_threshold=37
)

num_memories = 10000

# Create arrays for storing addresses and memories. 
addresses = np.empty((num_memories, SDM.address_dimension))
memories = np.empty((num_memories, SDM.memory_dimension))
recalled_memories = np.empty((num_memories, SDM.memory_dimension))
errors = np.empty(num_memories)

# Write all memories. 
start = time.time()
for i in tqdm(range(num_memories)): 
    address = np.random.randint(0, 2, SDM.address_dimension)
    memory = np.random.randint(0, 2, SDM.memory_dimension)
    addresses[i] = address
    memories[i] = memory
    SDM.write(address, memory)
    
# Read all memories.  
for i in tqdm(range(num_memories)): 
    address = addresses[i]
    recalled_memory = SDM.read(address)
    recalled_memories[i] = recalled_memory

end = time.time()

time.sleep(0.1)
os.system('cls' if os.name == 'nt' else 'clear')
print(f"\n\n\n\n\nElapsed time: {end-start:.5f}s\n\n\n\n\n")